internal/api/helpers.go

```diff
@@ -232,3 +233,18 @@ func isStringInSlice(checkValue string, list []string) bool {
 func getBodyBytes(req *http.Request) ([]byte, error) {
 	return utilities.GetBodyBytes(req)
 }
+
+func isValidCodeChallenge(codeChallenge string) bool {
+	// See Section 4.2: https://www.rfc-editor.org/rfc/rfc7636#section-4.2
+	hasValidChallengeChars := regexp.MustCompile("^[a-zA-Z-._~0-9]+$").MatchString
```

Comment:
```suggestion

var codeChallengePattern = regexp.MustCompile("^[a-zA-Z-._~0-9]+$")
func isValidCodeChallenge(codeChallenge string) bool {
	// See RFC 7636 Section 4.2: https://www.rfc-editor.org/rfc/rfc7636#section-4.2
	hasValidChallengeChars := codeChallengePattern.MatchString
```

Make sure you always extract regexp patterns to global variables, as each compilation is quite computationally expensive. Pattern objects are thread safe in Go, so it's fine to use them like this.